src/main/java/pl/allegro/tech/opel/OpelParser.java

```diff
@@ -22,24 +22,27 @@ Rule ParsingUnit() {
         return Sequence(WhiteSpace(), Program(), EOI);
     }
 
+    Rule Program() {
+        return Sequence(Declarations(), Expression(), Optional("; "), push(nodeFactory.program(pop(1), pop())));
+    }
+
     Rule Factor() {
         return FirstOf(
                 ifExpression(),
-                Sequence(
-                        FunctionCall(),
-                        ZeroOrMore(FirstOf(MethodCall(), ZeroArgumentMethodCall(), FieldAccess()))),
-                Sequence(
-                        StringLiteral(),
-                        ZeroOrMore(FirstOf(MethodCall(), ZeroArgumentMethodCall()))),
-                Sequence(
-                        NamedValue(),
-                        ZeroOrMore(FirstOf(MethodCall(), ZeroArgumentMethodCall(), FieldAccess()))),
+                Sequence(Object(), Train()),
                 Number(),
-                NegativeNumber(),
-                Sequence("( ", Expression(), ") ")
+                NegativeNumber()
         );
     }
 
+    Rule Object() {
+        return FirstOf(FunctionCall(),StringLiteral(), NamedValue(), ListInstantiation(), Sequence("( ", Expression(), ") "));
```

Comment:
One change here: we can access field on string literal, but this doesn't have to be detected by parsing engine, it can be detected later